File: test-video-processing.ts
#!/usr/bin/env node
/**
 * Test script to process a YouTube video locally
 * - Downloads video audio
 * - Transcribes with AssemblyAI
 * - Creates audio segments
 * - Uploads segments to Cloudflare R2
 * - Stores metadata in Cloudflare D1
 */

import { exec } from 'child_process';
import { promisify } from 'util';
import { readFileSync, writeFileSync, unlinkSync } from 'fs';
import { join } from 'path';
import FormData from 'form-data';
import fetch from 'node-fetch';

const execAsync = promisify(exec);

const YOUTUBE_URL = 'https://www.youtube.com/watch?v=u9sU5l92Th4';
const VIDEO_ID = 'u9sU5l92Th4';

// API Keys
const API_KEYS = {
  assemblyai: '4c15846aff03429e99207a86450addae',
  elevenlabs: 'sk_b3f632622b08afb9a26b2fb912be9d1baa2548414f430543',
};

// Cloudflare Worker URL (use local dev server or deployed)
const CLOUDFLARE_WORKER_URL = process.env.CLOUDFLARE_WORKER_URL || 'http://localhost:8787';

console.log('🎬 Starting local video processing test...\n');
console.log(`📺 Video URL: ${YOUTUBE_URL}`);
console.log(`📋 Video ID: ${VIDEO_ID}\n`);

async function downloadVideoAudio(videoId: string): Promise<string> {
  console.log('📥 Step 1: Downloading video audio...');
  
  const outputPath = join(process.cwd(), 'temp', `${videoId}.mp3`);
  
  try {
    // Create temp directory if it doesn't exist
    await execAsync(`mkdir -p temp`);
    
    // Download audio using yt-dlp
    const cmd = `yt-dlp --extract-audio --audio-format mp3 --output "${outputPath}" "${YOUTUBE_URL}"`;
    await execAsync(cmd, { timeout: 300000 }); // 5 minute timeout
    
    console.log(`✅ Audio downloaded: ${outputPath}`);
    return outputPath;
  } catch (error: any) {
    console.error('❌ Download failed:', error.message);
    throw error;
  }
}

async function transcribeWithAssemblyAI(youtubeUrl: string): Promise<{ text: string; words: Array<{ start: number; end: number; text: string }> }> {
  console.log('\n🎤 Step 2: Transcribing with AssemblyAI...');
  
  // Start transcription
  const startResponse = await fetch('https://api.assemblyai.com/v2/transcript', {
    method: 'POST',
    headers: {
      'Authorization': API_KEYS.assemblyai,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      audio_url: youtubeUrl,
      language_code: 'ps', // Pashto
      language_detection: true,
      word_boost: ['خدای', 'عیسی', 'پیغمبر', 'کتاب', 'تورات', 'انجیل'],
    }),
  });

  if (!startResponse.ok) {
    const errorText = await startResponse.text();
    throw new Error(`AssemblyAI start error: ${startResponse.status} - ${errorText}`);
  }

  const job = await startResponse.json();
  const transcriptId = job.id;
  console.log(`   Transcription job started: ${transcriptId}`);

  // Poll for completion
  let attempt = 0;
  const maxAttempts = 120; // 10 minutes max
  
  while (attempt < maxAttempts) {
    await new Promise(resolve => setTimeout(resolve, 5000)); // Wait 5 seconds
    
    const statusResponse = await fetch(`https://api.assemblyai.com/v2/transcript/${transcriptId}`, {
      headers: { 'Authorization': API_KEYS.assemblyai },
    });

    if (!statusResponse.ok) {
      throw new Error(`AssemblyAI status error: ${statusResponse.status}`);
    }

    const status = await statusResponse.json();
    
    if (status.status === 'completed') {
      console.log('✅ Transcription completed!');
      console.log(`   Transcript preview: ${status.text.substring(0, 100)}...`);
      return {
        text: status.text,
        words: status.words || [],
      };
    } else if (status.status === 'error') {
      throw new Error(`AssemblyAI error: ${status.error}`);
    }

    attempt++;
    if (attempt % 12 === 0) {
      console.log(`   ⏳ Waiting... (${Math.floor(attempt * 5 / 60)} minutes elapsed)`);
    }
  }

  throw new Error('Transcription timeout after 10 minutes');
}

function segmentTranscript(
  words: Array<{ start: number; end: number; text: string }>
): Array<{ text: string; startTime: number; endTime: number }> {
  console.log('\n✂️ Step 3: Segmenting transcript...');
  
  const segments: Array<{ text: string; startTime: number; endTime: number }> = [];
  let currentSegment: typeof words = [];

  for (const word of words) {
    currentSegment.push(word);
    
    const endOfSentence = word.text.endsWith('.') || word.text.endsWith('؟') || word.text.endsWith('!');
    const tooManyWords = currentSegment.length >= 15;

    if ((endOfSentence && currentSegment.length >= 3) || (tooManyWords && currentSegment.length >= 10)) {
      if (currentSegment.length > 0) {
        const text = currentSegment.map(w => w.text).join(' ');
        const startTime = currentSegment[0].start / 1000;
        const endTime = currentSegment[currentSegment.length - 1].end / 1000;

        segments.push({ text, startTime, endTime });
        currentSegment = [];
      }
    }
  }

  // Add remaining words
  if (currentSegment.length > 0) {
    const text = currentSegment.map(w => w.text).join(' ');
    const startTime = currentSegment[0].start / 1000;
    const endTime = currentSegment[currentSegment.length - 1].end / 1000;
    segments.push({ text, startTime, endTime });
  }

  console.log(`✅ Created ${segments.length} segments`);
  return segments;
}

async function extractAudioSegments(
  audioFile: string,
  segments: Array<{ startTime: number; endTime: number }>,
  videoId: string
): Promise<string[]> {
  console.log('\n🎵 Step 4: Extracting audio segments...');
  
  const segmentFiles: string[] = [];
  
  for (let i = 0; i < segments.length; i++) {
    const segment = segments[i];
    const outputPath = join(process.cwd(), 'temp', `${videoId}_segment_${i + 1}.mp3`);
    
    try {
      // Extract segment using ffmpeg
      const start = Math.floor(segment.startTime);
      const duration = Math.ceil(segment.endTime - segment.startTime);
      
      const cmd = `ffmpeg -i "${audioFile}" -ss ${start} -t ${duration} -acodec copy "${outputPath}" -y`;
      await execAsync(cmd, { timeout: 60000 });
      
      segmentFiles.push(outputPath);
      console.log(`   ✓ Segment ${i + 1}/${segments.length}: ${start}s - ${start + duration}s`);
    } catch (error: any) {
      console.error(`   ✗ Failed to extract segment ${i + 1}:`, error.message);
    }
  }

  console.log(`✅ Extracted ${segmentFiles.length} audio segments`);
  return segmentFiles;
}

async function uploadToCloudflare(segments: Array<{ text: string; startTime: number; endTime: number }>, segmentFiles: string[]): Promise<void> {
  console.log('\n☁️ Step 5: Uploading to Cloudflare...');
  
  // First, store metadata in D1 via Worker
  console.log('   Storing metadata in D1...');
  
  const metadataResponse = await fetch(`${CLOUDFLARE_WORKER_URL}/api/video/process`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      youtubeUrl: YOUTUBE_URL,
      videoId: VIDEO_ID,
      transcript: segments.map(s => s.text).join(' '),
      segments: segments,
      apiKeys: API_KEYS,
    }),
  });

  if (!metadataResponse.ok) {
    const errorText = await metadataResponse.text();
    console.error('❌ Failed to store metadata:', errorText);
    throw new Error(`Metadata storage failed: ${errorText}`);
  }

  const metadataResult = await metadataResponse.json();
  console.log('✅ Metadata stored in D1');

  // For R2 upload, we'd need to use wrangler R2 API or Cloudflare Worker
  // Since we're testing locally, let's use the Worker's R2 binding
  console.log('\n   Uploading audio clips to R2...');
  
  // Note: To upload to R2, we need to either:
  // 1. Use wrangler CLI: wrangler r2 object put videos/${VIDEO_ID}/segment_1.mp3 --file=temp/segment_1.mp3
  // 2. Or create an upload endpoint in the Worker
  
  // For now, let's show how to do it with wrangler CLI
  console.log('   📝 Note: Uploading via wrangler CLI...');
  
  for (let i = 0; i < segmentFiles.length; i++) {
    const segmentFile = segmentFiles[i];
    const r2Key = `videos/${VIDEO_ID}/segment_${i + 1}.mp3`;
    
    try {
      // Use wrangler to upload
      const cmd = `wrangler r2 object put "pashto-bible-audio/${r2Key}" --file="${segmentFile}"`;
      await execAsync(cmd);
      console.log(`   ✓ Uploaded segment ${i + 1} to R2: ${r2Key}`);
    } catch (error: any) {
      console.warn(`   ⚠ Failed to upload segment ${i + 1}:`, error.message);
      console.log(`   💡 Make sure wrangler is configured and R2 bucket exists`);
    }
  }

  console.log('✅ Upload complete!');
}

async function cleanup(files: string[]): Promise<void> {
  console.log('\n🧹 Cleaning up temporary files...');
  
  for (const file of files) {
    try {
      unlinkSync(file);
      console.log(`   ✓ Deleted ${file}`);
    } catch (error) {
      // Ignore cleanup errors
    }
  }
}

async function main() {
  try {
    const audioFile = await downloadVideoAudio(VIDEO_ID);
    const transcription = await transcribeWithAssemblyAI(YOUTUBE_URL);
    const segments = segmentTranscript(transcription.words);
    const segmentFiles = await extractAudioSegments(audioFile, segments, VIDEO_ID);
    
    await uploadToCloudflare(segments, segmentFiles);
    
    console.log('\n✅ Video processing complete!');
    console.log(`\n📊 Summary:`);
    console.log(`   Video ID: ${VIDEO_ID}`);
    console.log(`   Transcript: ${transcription.text.substring(0, 100)}...`);
    console.log(`   Segments: ${segments.length}`);
    console.log(`   Audio clips: ${segmentFiles.length}`);
    console.log(`\n🌐 View in Videos tab or check Cloudflare D1/R2`);
    
    // Cleanup
    await cleanup([audioFile, ...segmentFiles]);
    
  } catch (error: any) {
    console.error('\n❌ Processing failed:', error.message);
    process.exit(1);
  }
}

main();

File: test_video_processing.py
#!/usr/bin/env python3
"""
Test script to process YouTube video locally and upload to Cloudflare
- Downloads video audio
- Transcribes with AssemblyAI (via API)
- Extracts audio segments
- Uploads segments to Cloudflare R2
- Stores metadata in Cloudflare D1
"""

import os
import sys
import json
import time
import requests
import subprocess
from pathlib import Path
from typing import List, Dict

YOUTUBE_URL = 'https://www.youtube.com/watch?v=u9sU5l92Th4'
VIDEO_ID = 'u9sU5l92Th4'
API_URL = 'http://localhost:3000'
CLOUDFLARE_WORKER_URL = os.getenv('CLOUDFLARE_WORKER_URL', 'https://pashtobiblesearch.jeremy-samuels17.workers.dev')

API_KEYS = {
    'elevenlabs': 'sk_b3f632622b08afb9a26b2fb912be9d1baa2548414f430543',
    'assemblyai': '4c15846aff03429e99207a86450addae',
    'huggingface': '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',
    'deepseek': 'sk-9d567276d4ad41a08a074a0a83de1a67',
}

def download_video_audio(video_id: str) -> str:
    """Download video audio using yt-dlp"""
    print('📥 Step 1: Downloading video audio...')
    
    temp_dir = Path('temp')
    temp_dir.mkdir(exist_ok=True)
    output_path = temp_dir / f'{video_id}.mp3'
    
    try:
        cmd = [
            'yt-dlp',
            '--extract-audio',
            '--audio-format', 'mp3',
            '--output', str(output_path),
            YOUTUBE_URL
        ]
        
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=300)
        if result.returncode != 0:
            raise Exception(f'yt-dlp failed: {result.stderr}')
        
        print(f'✅ Audio downloaded: {output_path}')
        return str(output_path)
    except Exception as e:
        print(f'❌ Download failed: {e}')
        raise

def transcribe_with_assemblyai(audio_file: str) -> Dict:
    """Transcribe audio file using AssemblyAI"""
    print('\n🎤 Step 2: Transcribing with AssemblyAI...')
    
    # First, upload audio file to AssemblyAI
    print('   Uploading audio file...')
    with open(audio_file, 'rb') as f:
        upload_response = requests.post(
            'https://api.assemblyai.com/v2/upload',
            headers={'Authorization': API_KEYS['assemblyai']},
            files={'file': f},
            timeout=300
        )
    
    if not upload_response.ok:
        raise Exception(f'AssemblyAI upload error: {upload_response.status_code} - {upload_response.text}')
    
    upload_url = upload_response.json()['upload_url']
    print(f'   Audio uploaded: {upload_url}')
    
    # Start transcription
    response = requests.post(
        'https://api.assemblyai.com/v2/transcript',
        headers={
            'Authorization': API_KEYS['assemblyai'],
            'Content-Type': 'application/json',
        },
        json={
            'audio_url': upload_url,
            'language_code': 'ps',
            'word_boost': ['خدای', 'عیسی', 'پیغمبر', 'کتاب', 'تورات', 'انجیل'],
        },
        timeout=30
    )
    
    if not response.ok:
        raise Exception(f'AssemblyAI start error: {response.status_code} - {response.text}')
    
    job = response.json()
    transcript_id = job['id']
    print(f'   Transcription job started: {transcript_id}')
    
    # Poll for completion
    attempt = 0
    max_attempts = 120
    
    while attempt < max_attempts:
        time.sleep(5)
        
        status_response = requests.get(
            f'https://api.assemblyai.com/v2/transcript/{transcript_id}',
            headers={'Authorization': API_KEYS['assemblyai']},
            timeout=30
        )
        
        if not status_response.ok:
            raise Exception(f'AssemblyAI status error: {status_response.status_code}')
        
        status = status_response.json()
        
        if status['status'] == 'completed':
            print('✅ Transcription completed!')
            print(f'   Transcript preview: {status["text"][:100]}...')
            return {
                'text': status['text'],
                'words': status.get('words', []),
            }
        elif status['status'] == 'error':
            raise Exception(f'AssemblyAI error: {status.get("error", "Unknown error")}')
        
        attempt += 1
        if attempt % 12 == 0:
            print(f'   ⏳ Waiting... ({attempt * 5 // 60} minutes elapsed)')
    
    raise Exception('Transcription timeout after 10 minutes')

def segment_transcript(words: List[Dict]) -> List[Dict]:
    """Segment transcript into clips"""
    print('\n✂️ Step 3: Segmenting transcript...')
    
    segments = []
    current_segment = []
    
    for word in words:
        current_segment.append(word)
        
        end_of_sentence = word['text'].endswith('.') or word['text'].endswith('؟') or word['text'].endswith('!')
        too_many_words = len(current_segment) >= 15
        
        if (end_of_sentence and len(current_segment) >= 3) or (too_many_words and len(current_segment) >= 10):
            if current_segment:
                text = ' '.join(w['text'] for w in current_segment)
                start_time = current_segment[0]['start'] / 1000
                end_time = current_segment[-1]['end'] / 1000
                
                segments.append({
                    'text': text,
                    'startTime': start_time,
                    'endTime': end_time,
                })
                current_segment = []
    
    # Add remaining words
    if current_segment:
        text = ' '.join(w['text'] for w in current_segment)
        start_time = current_segment[0]['start'] / 1000
        end_time = current_segment[-1]['end'] / 1000
        segments.append({
            'text': text,
            'startTime': start_time,
            'endTime': end_time,
        })
    
    print(f'✅ Created {len(segments)} segments')
    return segments

def extract_audio_segments(audio_file: str, segments: List[Dict], video_id: str) -> List[str]:
    """Extract audio segments using ffmpeg"""
    print('\n🎵 Step 4: Extracting audio segments...')
    
    segment_files = []
    temp_dir = Path('temp')
    
    for i, segment in enumerate(segments):
        output_path = temp_dir / f'{video_id}_segment_{i + 1}.mp3'
        
        try:
            start = int(segment['startTime'])
            duration = int(segment['endTime'] - segment['startTime'])
            
            cmd = [
                'ffmpeg',
                '-i', audio_file,
                '-ss', str(start),
                '-t', str(duration),
                '-acodec', 'copy',
                str(output_path),
                '-y'
            ]
            
            result = subprocess.run(cmd, capture_output=True, text=True, timeout=60)
            if result.returncode != 0:
                print(f'   ⚠ Failed to extract segment {i + 1}: {result.stderr[:100]}')
                continue
            
            segment_files.append(str(output_path))
            print(f'   ✓ Segment {i + 1}/{len(segments)}: {start}s - {start + duration}s')
        except Exception as e:
            print(f'   ✗ Failed to extract segment {i + 1}: {e}')
    
    print(f'✅ Extracted {len(segment_files)} audio segments')
    return segment_files

def upload_to_r2(segment_files: List[str], video_id: str):
    """Upload audio segments to Cloudflare R2 using wrangler"""
    print('\n☁️ Step 5: Uploading to Cloudflare R2...')
    
    for i, segment_file in enumerate(segment_files):
        r2_key = f'videos/{video_id}/segment_{i + 1}.mp3'
        
        try:
            cmd = [
                'wrangler', 'r2', 'object', 'put',
                f'pashto-bible-audio/{r2_key}',
                '--file', segment_file
            ]
            
            result = subprocess.run(cmd, capture_output=True, text=True, timeout=60)
            if result.returncode != 0:
                print(f'   ⚠ Failed to upload segment {i + 1}: {result.stderr[:200]}')
                continue
            
            print(f'   ✓ Uploaded segment {i + 1} to R2: {r2_key}')
        except Exception as e:
            print(f'   ✗ Failed to upload segment {i + 1}: {e}')
    
    print('✅ Upload complete!')

def store_metadata_in_d1(segments: List[Dict], video_id: str):
    """Store metadata in Cloudflare D1 via Worker"""
    print('\n💾 Step 6: Storing metadata in Cloudflare D1...')
    
    try:
        # Combine transcript text
        transcript_text = ' '.join(s['text'] for s in segments)
        
        response = requests.post(
            f'{CLOUDFLARE_WORKER_URL}/api/video/process',
            headers={'Content-Type': 'application/json'},
            json={
                'youtubeUrl': YOUTUBE_URL,
                'videoId': video_id,
                'transcript': transcript_text,
                'segments': segments,
                'apiKeys': API_KEYS,
            },
            timeout=30
        )
        
        if response.ok:
            print('✅ Metadata stored in D1')
            result = response.json()
            print(f'   Video ID: {result.get("videoId", video_id)}')
            print(f'   Segments: {len(result.get("segments", []))}')
        else:
            print(f'⚠️ Failed to store metadata: {response.status_code} - {response.text[:200]}')
    except Exception as e:
        print(f'⚠️ Failed to store metadata: {e}')

def cleanup(files: List[str]):
    """Clean up temporary files"""
    print('\n🧹 Cleaning up temporary files...')
    
    for file_path in files:
        try:
            Path(file_path).unlink(missing_ok=True)
            print(f'   ✓ Deleted {file_path}')
        except Exception:
            pass

def main():
    print('🎬 Starting local video processing test...\n')
    print(f'📺 Video URL: {YOUTUBE_URL}\n')
    
    segment_files = []
    audio_file = None
    
    try:
        # Step 1: Download audio
        audio_file = download_video_audio(VIDEO_ID)
        
        # Step 2: Transcribe
        transcription = transcribe_with_assemblyai(audio_file)
        
        # Step 3: Segment
        segments = segment_transcript(transcription['words'])
        
        # Step 4: Extract segments
        segment_files = extract_audio_segments(audio_file, segments, VIDEO_ID)
        
        # Step 5: Upload to R2
        if segment_files:
            upload_to_r2(segment_files, VIDEO_ID)
        
        # Step 6: Store metadata in D1
        store_metadata_in_d1(segments, VIDEO_ID)
        
        # Summary
        print('\n✅ Video processing complete!')
        print(f'\n📊 Summary:')
        print(f'   Video ID: {VIDEO_ID}')
        print(f'   Transcript preview: {transcription["text"][:200]}...')
        print(f'   Segments: {len(segments)}')
        print(f'   Audio clips: {len(segment_files)}')
        print(f'\n🌐 Metadata stored in Cloudflare D1')
        print(f'📁 Audio clips stored in Cloudflare R2')
        
    except Exception as e:
        print(f'\n❌ Processing failed: {e}')
        import traceback
        traceback.print_exc()
        sys.exit(1)
    finally:
        # Cleanup
        if audio_file:
            cleanup([audio_file] + segment_files)

if __name__ == '__main__':
    main()

